File: output_20260318_123118/gpt52/java/Linklist/Spouse.java
class Person {

    protected Person spouse;

    /*@ predicate person(Person p; Person s) =
            p.spouse |-> s;
    @*/

    public void spouse_symm()
        //@ requires person(this, ?s);
        //@ ensures person(this, s);
    {
    }

    public Person()
        //@ requires true;
        //@ ensures person(this, null);
    {
    }
    
    public Person getSpouse()
        //@ requires person(this, ?s);
        //@ ensures person(this, s) &*& result == s;
    {
        return spouse;
    }
    
    protected void setSpouse(Person other)
        //@ requires person(this, ?s0) &*& person(other, ?s1);
        //@ ensures person(this, other) &*& person(other, this);
    {
        spouse = other;
        other.spouse = this;
    }
    
    protected void clearSpouse()
        //@ requires person(this, ?s) &*& s != null &*& person(s, this);
        //@ ensures person(this, null) &*& person(s, null);
    {
        spouse.spouse = null;
        spouse = null;
    }
    
    void marry(Person other)
        //@ requires person(this, ?s0) &*& person(other, ?s1);
        //@ ensures person(this, other) &*& person(other, this);
    {
        other.setSpouse(this);
    }
    
    void divorce()
        //@ requires person(this, ?s) &*& s != null &*& person(s, this);
        //@ ensures person(this, null) &*& person(s, null);
    {
        spouse.clearSpouse();
    }

}

class Program {

    public static void main(String[] args)
        //@ requires true;
        //@ ensures true;
    {
        Person a = new Person();
        Person b = new Person();
        //@ open person(a, null);
        //@ open person(b, null);
        a.marry(b);
        //@ open person(a, b);
        //@ open person(b, a);
        b.divorce();
        //@ open person(a, null);
        //@ open person(b, null);
    }

}

File: output_20260321_024012/claude-opus/c/Struct/spouse.c
#include "stdlib.h"
#include "spouse.h"

/*@
predicate person(struct person* p;)
/*@  p->name |-> ?name &*& p->spouse |-> ?spouse; @*/
@*/

struct person {
  char* name;
  struct person* spouse;
};

struct person *create_person()
  
//@ requires true;
//@ ensures person(result);
{
  struct person *p = malloc(sizeof(struct person));
  if(p == 0) abort();
  //@ close person(p);
  p->spouse = 0;
  
  return p;
}

void marry(struct person *this, struct person *other)
  
//@ requires person(this) &*& person(other);
//@ ensures person(this) &*& person(other);
{
  //@ open person(this);
  //@ open person(other);
  this->spouse = other;
  other->spouse = this;
  //@ close person(this);
  //@ close person(other);
}

struct person* person_get_spouse(struct person* this)
  
//@ requires person(this);
//@ ensures person(this) &*& (result == 0 || person(result));
{
  //@ open person(this);
  struct person* res = this->spouse;
  //@ if(res != 0) { open person(res); close person(res); }
  //@ close person(this);
  return res;
}

void divorce(struct person* this)
  
//@ requires person(this) &*& this->spouse != 0 &*& person(this->spouse);
//@ ensures person(this) &*& person(this->spouse);
{
  //@ open person(this);
  //@ open person(this->spouse);
  this->spouse->spouse = 0;
  this->spouse = 0;
  //@ close person(this->spouse);
  //@ close person(this);
}

void die(struct person *this)
  
//@ requires person(this);
//@ ensures true;
{
  //@ open person(this);
  if(this->spouse != 0) {
    //@ open person(this->spouse);
    this->spouse->spouse = 0;
    //@ close person(this->spouse);
  }
  free(this); 
}

int main() 
  
//@ requires true;
//@ ensures true;
{
  struct person* alice = create_person();
  struct person* bob = create_person();
  struct person* eve = 0;
  marry(alice, bob);
  eve = create_person();
  divorce(bob);
  marry(bob, eve);
  die(alice);
  die(bob);
  die(eve);
  return 0;
}